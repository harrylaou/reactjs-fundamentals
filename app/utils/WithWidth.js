import React, {Component} from 'react';

export const SMALL = 1;
export const MEDIUM = 2;
export const LARGE = 3;

export default function withWidth(options = {}) {
    const {
        largeWidth = 992,
        mediumWidth = 768,
        resizeInterval = 166,
    } = options;

    return (MyComponent) => {

        return class WithWidth extends Component {

            constructor() {
                super();
                this.state = { width: this.windowWidth()};
                this.handleResize = this.handleResize.bind(this);
            }

            componentDidMount() {
                if (window) {
                    window.addEventListener('resize', this.handleResize);
                    this.handleResize();
                }
            }

            componentWillUnmount() {
                if (window) {
                    window.removeEventListener('resize', this.handleResize);
                }
            }

            handleResize() {
                let width = this.windowWidth()
                if (width !== this.state.width) {
                    this.setState({
                        width: width,
                    });
                }
            };

            windowWidth() {
                let innerWidth = 0;
                let width;
                if (window) {
                    innerWidth = window.innerWidth;
                }
                if (innerWidth >= largeWidth) {
                width = LARGE;
                } else if (innerWidth >= mediumWidth) {
                width = MEDIUM;
                } else { // innerWidth < 768
                width = SMALL;
                }
                return width;
            }

            render() {
                return (
                    <MyComponent
                    {...this.props}
                    width={this.state.width}
                    />
                );
            }
        };
    };
}
