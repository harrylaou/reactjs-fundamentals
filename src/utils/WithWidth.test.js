import React from 'react'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import WithWidth from './WithWidth'

describe('<WithWidth />', () => {
  const DemoElem = () => (
    <div>Hello World</div>
  )

  const Elem = WithWidth(DemoElem)

  it('should render with default props', () => {

    const wrapper = shallow(<Elem />)
    expect(shallowToJson(wrapper.dive())).toMatchSnapshot()
  })

  it('should get LARGE width with default props', () => {

    global.innerWidth = 1024
    const wrapper = shallow(<Elem />)
    expect(wrapper.prop('width')).toEqual(3)
  })

  it('should get MEDIUM width with default props', () => {

    global.innerWidth = 768
    const wrapper = shallow(<Elem />)
    expect(wrapper.prop('width')).toEqual(2)
  })

  it('should get SMALL width with default props', () => {

    global.innerWidth = 166
    const wrapper = shallow(<Elem />)
    expect(wrapper.prop('width')).toEqual(1)
  })
})