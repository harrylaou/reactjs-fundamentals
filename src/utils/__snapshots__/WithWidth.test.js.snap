// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WithWidth /> should render with default props 1`] = `
<div>
  Hello World
</div>
`;
